src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,523 @@
+        /// <returns>
+        /// The amount of data actually returned. If this size matches the bufferSize parameter then
+        /// the data is valid. If the sizes do not match then the data is invalid, most likely due 
+        /// to not having enough permissions to query for the data of that specific process
+        /// </returns>
+        [DllImport(Interop.Libraries.libproc, SetLastError = true)]
+        private static extern int proc_pidinfo(
+            int pid,
+            int flavor,
+            ulong arg,
+            IntPtr buffer,
+            int bufferSize);
+
+        /// <summary>
+        /// Gets the process information for a given process
+        /// </summary>
+        /// <param name="pid">The PID (process ID) of the process</param>
+        /// <returns>
+        /// Returns a valid proc_taskallinfo struct for valid processes that the caller 
+        /// has permission to access; otherwise, returns null
+        /// </returns>
+        internal static proc_taskallinfo? GetProcessInfoById(int pid)
+        {
+            proc_taskallinfo info;
+            bool success = internal_proc_pidinfo_helper<proc_taskallinfo>(pid, PROC_PIDTASKALLINFO, 0, out info);
+            return (success ? new proc_taskallinfo?(info) : null);
+        }
+
+        /// <summary>
+        /// Gets the thread information for the given thread
+        /// </summary>
+        /// <param name="thread">The ID of the thread to query for information</param>
+        /// <returns>
+        /// Returns a valid proc_threadinfo struct for valid threads that the caller
+        /// has permissions to access; otherwise, returns null
+        /// </returns>
+        internal static proc_threadinfo? GetThreadInfoById(ulong thread)
+        {
+            proc_threadinfo info;
+            bool success = internal_proc_pidinfo_helper<proc_threadinfo>(Interop.libc.getpid(), PROC_PIDTHREADINFO, (ulong)thread, out info);
+            return (success ? new proc_threadinfo?(info) : null);
+        }
+
+        internal static List<KeyValuePair<ulong, proc_threadinfo?>> GetAllThreadsInProcess(int pid)
+        {
+            // Allocate unmanaged memory for the thread IDs
+            int size = PROC_PIDLISTTHREADS_SIZE * 20; // start assuming 20 threads is enough
+            IntPtr buffer = Marshal.AllocCoTaskMem(size);
+
+            // Try to get the thread IDs
+            int result = proc_pidinfo(pid, PROC_PIDLISTTHREADS, 0,  buffer, size);
+
+            // Since we don't know how many threads there could be, if result == size, that could mean two things
+            // 1) We guessed exactly how many threads there are
+            // 2) There are more threads that we didn't get since our buffer is too small
+            // To make sure it isn't #2, when we result == size, double the buffer and try again
+            while (result == size)
+            {
+                Marshal.FreeCoTaskMem(buffer);
```

Comment:
NIT: Would probably use AllocHGlobal and friends here, since you don't actually care about COM (I doubt there's a difference on CoreCLR XPlat, so this is maybe too pedantic).